code/game/machinery/doors/firedoor.dm

```diff
@@ -23,12 +23,86 @@
 	var/nextstate = null
 	var/boltslocked = TRUE
-	var/list/affecting_areas
 	var/being_held_open = FALSE
+	/// The merger id used to create/get the merger group in charge of handling tanks that share an internal gas storage
+	var/merger_id = "firelock"
+	///Don't register the turfs signal more than once
+	var/registered_turfs = FALSE
+	///Stops the firealarm from reopening the door if there is an hazard nearby
+	var/hazard = FALSE
 
 /obj/machinery/door/firedoor/Initialize()
 	. = ..()
-	CalculateAffectingAreas()
+	var/area/my_area = get_area(src)
+	LAZYADD(my_area.firedoors, src)
+
+	RegisterSignal(src, COMSIG_MERGER_ADDING, .proc/MergerAdding)
+	RegisterSignal(src, COMSIG_MERGER_REMOVING, .proc/MergerRemoving)
+
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/machinery/door/firedoor/LateInitialize()
+	. = ..()
+	GetMergeGroup(merger_id, allowed_types = typecacheof(/obj/machinery/door/firedoor))
+
+/obj/machinery/door/firedoor/Destroy()
+	var/area/my_area = get_area(src)
+	LAZYREMOVE(my_area.firedoors,src)
+	return..()
+
+///////////////////////////////////////////////////////////////////
+// Merger handling
+
+/obj/machinery/door/firedoor/proc/MergerAdding(obj/machinery/door/firedoor/us, datum/merger/new_merger)
+	SIGNAL_HANDLER
+	if(new_merger.id != merger_id)
+		return
+	RegisterSignal(new_merger, COMSIG_MERGER_REFRESH_COMPLETE, .proc/MergerRefreshComplete)
+
+/obj/machinery/door/firedoor/proc/MergerRemoving(obj/machinery/door/firedoor/us, datum/merger/old_merger)
+	SIGNAL_HANDLER
+	if(old_merger.id != merger_id)
+		return
+	UnregisterSignal(old_merger, COMSIG_MERGER_REFRESH_COMPLETE)
+
+/// Handles the firelocks to register only one signal per group
+/obj/machinery/door/firedoor/proc/MergerRefreshComplete(datum/merger/merger, list/leaving_members, list/joining_members)
+	SIGNAL_HANDLER
+	var/turf/open/turf1 = get_step(src, dir)
+	var/turf/open/turf2 = get_step(src, turn(dir, 180))
+	if(merger.origin != src)
+		registered_turfs = FALSE
+		UnregisterSignal(turf1, COMSIG_TURF_EXPOSE)
+		UnregisterSignal(turf2, COMSIG_TURF_EXPOSE)
+		return
+
+	if(!registered_turfs)
+		RegisterSignal(turf1, COMSIG_TURF_EXPOSE, .proc/check_closing)
+		RegisterSignal(turf2, COMSIG_TURF_EXPOSE, .proc/check_closing)
+		registered_turfs = TRUE
+
+///Check if the temperature and pressure changes are enough to cause body damage
+/obj/machinery/door/firedoor/proc/check_closing(datum/source, datum/gas_mixture/environment, exposed_temperature)
+	SIGNAL_HANDLER
+	var/temperature_hazard = TRUE
+	if(exposed_temperature > BODYTEMP_COLD_DAMAGE_LIMIT && exposed_temperature < BODYTEMP_HEAT_DAMAGE_LIMIT)
+		temperature_hazard = FALSE
+
+	var/environment_pressure = environment.return_pressure()
+	var/pressure_hazard = TRUE
+	if(environment_pressure > HAZARD_LOW_PRESSURE && environment_pressure < HAZARD_HIGH_PRESSURE)
+		pressure_hazard = FALSE
```

Comment:
Undocced change?

Reply:
what you mean? i added in the PR text that firedoors can close due to low/high pressure

Reply:
Oh you did. I'm blind